src/OSDP.Net/ControlPanel.cs

```diff
@@ -71,54 +71,76 @@ public ControlPanel(ILogger<ControlPanel> logger = null)
         /// <returns>An identifier that represents the connection</returns>
+        /// <exception cref="InvalidOperationException">If the connection is already started.</exception>
         public Guid StartConnection(IOsdpConnection connection, TimeSpan pollInterval, Action<TraceEntry> tracer)
         {
-            var existingBusWithThisConnection = _buses.Values
-                .FirstOrDefault(bus => bus.Connection == connection);
-            if(existingBusWithThisConnection != null)
-                throw new InvalidOperationException(
-                    $"The IOsdpConnection is already active in connection {existingBusWithThisConnection.Id}. " +
-                    "That connection must be stopped before starting a new one.");
-
-            var newBus = new Bus(
-                connection,
-                _replies,
-                pollInterval,
-                tracer,
-                _logger);
-
+            var newBus = CreateBus(connection, pollInterval, tracer);
             newBus.ConnectionStatusChanged += BusOnConnectionStatusChanged;
-
-            _buses[newBus.Id] = newBus;
-
             newBus.StartPolling();
-
             return newBus.Id;
         }
 
+        /// <summary>
+        /// Create a bus for a connection.
+        /// </summary>
+        /// <remarks>
+        /// This will serialize concurrent attempts to start simultaneous connections. Only the first will succeed.
+        /// </remarks>
+        private Bus CreateBus(IOsdpConnection connection, TimeSpan pollInterval, Action<TraceEntry> tracer)
+        {
+            // Lock while we check/create the bus. This is a very quick operation so the lock is not a performance problem.
+            lock (_buses)
+            {
+                var existingBusWithThisConnection = _buses.Values
+                    .FirstOrDefault(b => b.Connection == connection);
+                if (existingBusWithThisConnection != null)
+                    throw new InvalidOperationException(
+                        $"The IOsdpConnection is already active in connection {existingBusWithThisConnection.Id}. " +
+                        "That connection must be stopped before starting a new one.");
+
+                var newBus = new Bus(
+                    connection,
+                    _replies,
+                    pollInterval,
+                    tracer,
+                    _logger);
+
+                _buses[newBus.Id] = newBus;
+                return newBus;
+            }
+        }
+
         /// <summary>
         /// Stop the bus for a specific connection.
         /// </summary>
         /// <param name="connectionId">The identifier that represents the connection.</param>
         public async Task StopConnection(Guid connectionId)
         {
-            if (!_buses.TryRemove(connectionId, out Bus bus))
+            if (!_buses.TryGetValue(connectionId, out Bus bus))
```

Comment:
Keeping the bus in the collection until it is closed guarrantees it won't be started while it's closing by another thread.